File: frontend/src/api/onboarding.ts
/**
 * オンボーディングAPI
 */

import apiClient from './client';

// オンボーディングステップ名
export type OnboardingStepName =
  | 'welcome'
  | 'connect_platform'
  | 'select_goals'
  | 'setup_notifications'
  | 'first_analysis'
  | 'complete';

// ステップステータス
export type OnboardingStepStatus =
  | 'not_started'
  | 'in_progress'
  | 'completed'
  | 'skipped';

// オンボーディングステップ
export interface OnboardingStep {
  name: OnboardingStepName;
  status: OnboardingStepStatus;
  completed_at: string | null;
  data: Record<string, unknown> | null;
}

// オンボーディング状態レスポンス
export interface OnboardingStatusResponse {
  is_completed: boolean;
  current_step: OnboardingStepName;
  steps: OnboardingStep[];
  started_at: string | null;
  completed_at: string | null;
  progress_percent: number;
}

// ステップ完了リクエスト
export interface OnboardingCompleteStepRequest {
  step_name: OnboardingStepName;
  data?: Record<string, unknown>;
}

// スキップリクエスト
export interface OnboardingSkipRequest {
  reason?: string;
}

/**
 * オンボーディング状態を取得
 */
export async function getOnboardingStatus(): Promise<OnboardingStatusResponse> {
  const response = await apiClient.get<OnboardingStatusResponse>(
    '/api/v1/onboarding/status'
  );
  return response.data;
}

/**
 * オンボーディングを開始
 */
export async function startOnboarding(): Promise<OnboardingStatusResponse> {
  const response = await apiClient.post<OnboardingStatusResponse>(
    '/api/v1/onboarding/start'
  );
  return response.data;
}

/**
 * ステップを完了
 */
export async function completeOnboardingStep(
  request: OnboardingCompleteStepRequest
): Promise<OnboardingStatusResponse> {
  const response = await apiClient.post<OnboardingStatusResponse>(
    '/api/v1/onboarding/complete-step',
    request
  );
  return response.data;
}

/**
 * ステップをスキップ
 */
export async function skipOnboardingStep(
  stepName: OnboardingStepName
): Promise<OnboardingStatusResponse> {
  const response = await apiClient.post<OnboardingStatusResponse>(
    `/api/v1/onboarding/skip-step/${stepName}`
  );
  return response.data;
}

/**
 * オンボーディング全体をスキップ
 */
export async function skipAllOnboarding(
  request?: OnboardingSkipRequest
): Promise<OnboardingStatusResponse> {
  const response = await apiClient.post<OnboardingStatusResponse>(
    '/api/v1/onboarding/skip-all',
    request || {}
  );
  return response.data;
}

/**
 * オンボーディングをリセット（開発用）
 */
export async function resetOnboarding(): Promise<OnboardingStatusResponse> {
  const response = await apiClient.post<OnboardingStatusResponse>(
    '/api/v1/onboarding/reset'
  );
  return response.data;
}

File: src/api/routers/onboarding.py
"""
オンボーディングAPIルーター

新規ユーザーのオンボーディングフロー管理
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from ..db import get_db
from ..db.models import User
from ..dependencies import get_current_user
from ..onboarding import OnboardingService
from ..schemas import (
    OnboardingCompleteStepRequest,
    OnboardingSkipRequest,
    OnboardingStatusResponse,
)

router = APIRouter()


def get_onboarding_service(db: Session = Depends(get_db)) -> OnboardingService:
    """オンボーディングサービスを取得"""
    return OnboardingService(db)


@router.get(
    "/status",
    response_model=OnboardingStatusResponse,
    summary="オンボーディング状態取得",
    description="現在のユーザーのオンボーディング進捗状態を取得します",
)
def get_onboarding_status(
    current_user: User = Depends(get_current_user),
    service: OnboardingService = Depends(get_onboarding_service),
) -> OnboardingStatusResponse:
    """オンボーディング状態を取得"""
    return service.get_status(current_user)


@router.post(
    "/start",
    response_model=OnboardingStatusResponse,
    summary="オンボーディング開始",
    description="オンボーディングフローを開始します",
)
def start_onboarding(
    current_user: User = Depends(get_current_user),
    service: OnboardingService = Depends(get_onboarding_service),
) -> OnboardingStatusResponse:
    """オンボーディングを開始"""
    return service.start_onboarding(current_user)


@router.post(
    "/complete-step",
    response_model=OnboardingStatusResponse,
    summary="ステップ完了",
    description="指定したオンボーディングステップを完了します",
)
def complete_step(
    request: OnboardingCompleteStepRequest,
    current_user: User = Depends(get_current_user),
    service: OnboardingService = Depends(get_onboarding_service),
) -> OnboardingStatusResponse:
    """オンボーディングステップを完了"""
    return service.complete_step(
        current_user,
        request.step_name,
        request.data,
    )


@router.post(
    "/skip-step/{step_name}",
    response_model=OnboardingStatusResponse,
    summary="ステップスキップ",
    description="指定したオンボーディングステップをスキップします",
)
def skip_step(
    step_name: str,
    current_user: User = Depends(get_current_user),
    service: OnboardingService = Depends(get_onboarding_service),
) -> OnboardingStatusResponse:
    """オンボーディングステップをスキップ"""
    from ..schemas import OnboardingStepName

    try:
        step = OnboardingStepName(step_name)
    except ValueError:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"無効なステップ名: {step_name}",
        )

    try:
        return service.skip_step(current_user, step)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )


@router.post(
    "/skip-all",
    response_model=OnboardingStatusResponse,
    summary="オンボーディング全スキップ",
    description="オンボーディング全体をスキップして完了状態にします",
)
def skip_all_onboarding(
    request: OnboardingSkipRequest | None = None,
    current_user: User = Depends(get_current_user),
    service: OnboardingService = Depends(get_onboarding_service),
) -> OnboardingStatusResponse:
    """オンボーディング全体をスキップ"""
    return service.skip_all(current_user)


@router.post(
    "/reset",
    response_model=OnboardingStatusResponse,
    summary="オンボーディングリセット",
    description="オンボーディング状態をリセットします（開発/テスト用）",
)
def reset_onboarding(
    current_user: User = Depends(get_current_user),
    service: OnboardingService = Depends(get_onboarding_service),
) -> OnboardingStatusResponse:
    """オンボーディングをリセット"""
    return service.reset_onboarding(current_user)
